rust/codelist-builder-rs/src/snomed_usage_data.rs

```diff
@@ -0,0 +1,391 @@
+//! This file contains the snomed usage data struct and its implementation
+
+// Internal imports
+use crate::errors::CodeListBuilderError;
+use crate::usage_year::UsageYear;
+
+// External imports
+use csv;
+use reqwest;
+use serde::{Deserialize, Serialize};
+
+/// Struct to represent a snomed usage data entry
+///
+/// # Fields
+/// * `snomed_concept_id` - The snomed concept id
+/// * `description` - The description
+/// * `usage` - The usage
+/// * `active_at_start` - Whether the concept was active at the start of the usage period
+/// * `active_at_end` - Whether the concept was active at the end of the usage period
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SnomedUsageDataEntry {
+    pub snomed_concept_id: String,
+    pub description: String,
+    pub usage: String, // allows for * for count of 1-4
+    pub active_at_start: bool,
+    pub active_at_end: bool,
+}
+
+/// Struct to represent snomed usage data
+///
+/// # Fields
+/// * `usage_data` - The usage data
+/// * `usage_year` - The usage year
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SnomedUsageData {
+    pub usage_data: Vec<SnomedUsageDataEntry>,
+    pub usage_year: UsageYear,
+}
+
+impl SnomedUsageData {
+    /// Download snomed usage data from a url
+    ///
+    /// # Arguments
+    /// * `base_url` - The base url
+    /// * `usage_year` - The usage year
+    ///
+    /// # Returns
+    /// Self or an error if the download fails
+    pub async fn download_usage(
+        base_url: &str,
+        usage_year: UsageYear,
+    ) -> Result<Self, CodeListBuilderError> {
+        let url = format!(
+            "{}/{}",
+            base_url.trim_end_matches('/'),
+            usage_year.path().trim_start_matches('/')
+        );
+
+        let body = reqwest::get(&url).await?.text().await?;
+
+        let usage_data = Self::parse_from_string(&body)?;
```

Comment:
I don't think we need to do this now but if this file is big, we might want to stream in chunks to write to file.